lib/framereader.py

```diff
@@ -174,6 +174,34 @@ def read_frame_P410(fd, width, height):
 
   return y, u, v
 
+def read_frame_Y410(fd, width, height):
+  size = width * height * 4
+  #https://docs.microsoft.com/en-us/windows/win32/medfound/10-bit-and-16-bit-yuv-video-formats
+  #y410 32bit
+  #U bit[9:0]
+  #Y bit[19:10]
+  #V bit[29:20]
+  #A bit[31:30]
+  y410 = numpy.fromfile(fd, dtype=numpy.uint8, count=size)
+
+  y410_dw_4 = y410[3::4].reshape((height, width))
+  y410_dw_3 = y410[2::4].reshape((height, width))
+  y410_dw_2 = y410[1::4].reshape((height, width))
+  y410_dw_1 = y410[0::4].reshape((height, width))
+
+  y410_high16 = (uint16_t) (y410_dw_4<<8 | y410_dw_3)
+  y410_low16 = (uint16_t) (y410_dw_2<<8 | y410_dw_1)
+  #bit 20-30
+  v= (uint16_t) ( (y410_high16&0x3ff0) >> 4 )
+  #bit 10-19
+  y_h_4 = (uint16_t) (y410_high16 & 0x000f)
+  y_l_6 = (uint16_t) (y410_low16 & 0xfc00)
+  y = (uint16_t) ((u_h_4<<6) | (u_l_6>>10))
+  #bit 0-9
+  u = (uint16_t) (y410_low16 & 0x03ff)
+
```

Comment:
is it possible to take y410 as   an uint32_t ? then 
U = y410 & 0x3ff
Y = (y410 >> 10) & 0x3ff
V = (y410 >> 20) & 0x3ff

Reply:
Yes, it seems that your suggestion has more clean source codes

Reply:
What order are the bytes stored on disk/file?  If little-endian, then we can use `numpy.uint32`, otherwise we need to use `numpy.dtype(">u4")`.

This should be the most efficient (for LE file storage):
```python
  y410 = numpy.fromfile(fd, dtype=numpy.uint32, count=width*height)

  uf = lambda dw: dw & 0x3ff
  yf = lambda dw: (dw >> 10) & 0x3ff
  vf = lambda dw: (dw >> 20) & 0x3ff

  u = uf(y410).reshape((height, width))
  y = yf(y410).reshape((height, width))
  v = vf(y410).reshape((height, width))

  return y, u, v
```
For BE, just change to `dtype=numpy.dtype(">u4")`

... and here is a test (which we should add to  `test/self/formats.py`)
```python
@slash.parametrize(("width", "height"), itertools.product([175, 176, 177], [143, 144, 145]))
def test_frame_reader_Y410(width, height):
  testfile = get_media()._test_artifact(
    "test_frame_reader_Y410_{}x{}.Y410".format(width, height))
  a = 3; oy = 512; ou = 256; ov = 128;
  dword = (a << 30) | (ov << 20) | (oy << 10) | ou
  b = dword.to_bytes(4, "little")
  with open(testfile, "wb") as fd:
    for i in range(width * height):
      fd.write(b)

  with open(testfile, "rb") as fd:
    y, u, v = read_frame_Y410(fd, width, height)

    yf = lambda val: val == oy
    uf = lambda val: val == ou
    vf = lambda val: val == ov

    assert y.size == u.size == v.size == width * height
    assert fd.tell() == get_framesize(width, height, "Y410")
    assert yf(y).all()
    assert uf(u).all()
    assert vf(v).all()
```
For BE storage, just change to `dword.to_bytes(4, "big")`

Reply:
Hmmm, it seems we can apply operators directly to numpy.array types for even more efficiency/speed 😄:
```python
  y410 = numpy.fromfile(fd, dtype=numpy.uint32, count=width*height)
  u = (y410 & 0x3ff).reshape((height, width))
  y = ((y410 >> 10) & 0x3ff).reshape((height, width))
  v = ((y410 >> 20) & 0x3ff).reshape((height, width))
```

Reply:
thanks Artie, I am doing update based on your suggestion by using numpy.array types directly

Reply:
updated patch and added a unit test function -- test_frame_reader_Y410, all test cases passed

/home/focus/workspace_focus/hevc_10bit_yuv444/vaapi-fits# ./vaapi-fits run test/self/formats.py:test_frame_reader_Y410 --pl ICL
9 tests collected
test/self/formats.py .........  PASS
======================= Session ended. 9 successful, 0 skipped, 0 failed, 0 erroneous. Total duration: 00:00:03 =======================
root@media_libva_icl2:/home/focus/workspace_focus/hevc_10bit_yuv444/vaapi-fits#